cloudferrylib/os/actions/networks_transporter.py

```diff
@@ -13,16 +13,73 @@
 # limitations under the License.
 
 
+from fabric.api import settings
+
+from cloudferrylib.base.action import action
 from cloudferrylib.base.action import transporter
-from cloudferrylib.utils import utils as utl
+from cloudferrylib.utils import utils
+from cloudferrylib.utils import remote_runner
+
+
+NAMESPACE = 'routers'
 
 
 class NetworkTransporter(transporter.Transporter):
 
     def run(self, **kwargs):
-        src_resource = self.src_cloud.resources[utl.NETWORK_RESOURCE]
-        dst_resource = self.dst_cloud.resources[utl.NETWORK_RESOURCE]
+        src_resource = self.src_cloud.resources[utils.NETWORK_RESOURCE]
+        dst_resource = self.dst_cloud.resources[utils.NETWORK_RESOURCE]
         info = src_resource.read_info()
         dst_resource.deploy(info)
         return {}
 
+
+class ChangeRouterBase(action.Action):
+    def __init__(self, *args, **kwargs):
+        super(ChangeRouterBase, self).__init__(*args, **kwargs)
+        self.net = self.cloud.resources[utils.NETWORK_RESOURCE]
+        self.cfg = self.cloud.cloud_config.cloud
+
+    def modify_router(self, router):
+        node = self.net.get_router_host(router['id'])
+        runner = remote_runner.RemoteRunner(node, self.cfg.ssh_user,
+                                            password=self.cfg.ssh_sudo_password)
+        with settings(abort_on_prompts=True,
+                      gateway=self.cloud.getIpSsh()):
+            id = router['id']
+            _, ext_cidr_len = router['ext_cidr'].split('/')
+            int_cidr = router['int_cidr']
+            ext_ip = router['ext_ip']
+            new_ip = router['new_ip']
+            device = runner.run("ip netns exec qrouter-%s ip addr | grep %s"
+                                " | awk '{ print $NF }'" %
+                                (router['id'], router['ext_ip']))
+            runner.run("ip netns exec qrouter-%s iptables -D "
+                       "quantum-l3-agent-snat -s %s -t nat -j SNAT "
+                       "--to-source %s" % (id, int_cidr, ext_ip))
+            runner.run("ip netns exec qrouter-%s ip addr del %s/%s dev %s" %
+                       (id, ext_ip, ext_cidr_len, device))
+            runner.run("ip netns exec qrouter-%s ip addr add %s/%s dev %s" %
+                       (id, new_ip, ext_cidr_len, device))
+            runner.run("ip netns exec qrouter-%s iptables -A "
+                       "quantum-l3-agent-snat -s %s -t nat -j SNAT "
+                       "--to-source %s" % (id, int_cidr, new_ip))
+
+
+class ChangeRouter(ChangeRouterBase):
+    def run(self, **kwargs):
+        info = self.net.read_info()
+        routers_info = self.net.get_routers_for_changing(info)
+        for router in routers_info:
+            self.modify_router(router)
+        return {NAMESPACE: routers_info}
+
+
+class RollbackRouter(ChangeRouterBase):
```

Comment:
I'm assuming we must restore original state after migration has finished successfully too?.. In this case we must also add this task in the end of scenario, not only as rollback case.